src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp

```diff
@@ -15,29 +15,404 @@
+    ACTION_BERSERK
+};
+
+struct SummonLocation
+{
+    Position pos;
+    uint32 entry;
+};
+
+SummonLocation const PreAddLocations[] =
+{
+    { { 2149.68f, -263.477f, 419.679f, 3.120f }, NPC_JORMUNGAR_BEHEMOTH    },
+    { { 2131.31f, -271.640f, 419.840f, 2.188f }, NPC_MERCENARY_CAPTAIN_A   },
+    { { 2127.24f, -259.182f, 419.974f, 5.917f }, NPC_MERCENARY_SOLDIER_A   },
+    { { 2123.32f, -254.770f, 419.840f, 6.170f }, NPC_MERCENARY_SOLDIER_A   },
+    { { 2120.10f, -258.990f, 419.840f, 6.250f }, NPC_MERCENARY_SOLDIER_A   },
+    { { 2129.09f, -277.142f, 419.756f, 1.222f }, NPC_DARK_RUNE_ACOLYTE_PRE }
+};
+
+SummonLocation const ColossusAddLocations[] =
+{
+    { { 2218.38f, -297.50f, 412.18f, 1.030f }, NPC_IRON_RING_GUARD   },
+    { { 2235.07f, -297.98f, 412.18f, 1.613f }, NPC_IRON_RING_GUARD   },
+    { { 2235.26f, -338.34f, 412.18f, 1.589f }, NPC_IRON_RING_GUARD   },
+    { { 2217.69f, -337.39f, 412.18f, 1.241f }, NPC_IRON_RING_GUARD   },
+    { { 2227.58f, -308.30f, 412.18f, 1.591f }, NPC_DARK_RUNE_ACOLYTE },
+    { { 2227.47f, -345.37f, 412.18f, 1.566f }, NPC_DARK_RUNE_ACOLYTE }
+};
+
+SummonLocation const GiantAddLocations[] =
+{
+    { { 2198.05f, -428.77f, 419.95f, 6.056f }, NPC_IRON_HONOR_GUARD  },
+    { { 2220.31f, -436.22f, 412.26f, 1.064f }, NPC_IRON_HONOR_GUARD  },
+    { { 2158.88f, -441.73f, 438.25f, 0.127f }, NPC_IRON_HONOR_GUARD  },
+    { { 2198.29f, -436.92f, 419.95f, 0.261f }, NPC_DARK_RUNE_ACOLYTE },
+    { { 2230.93f, -434.27f, 412.26f, 1.931f }, NPC_DARK_RUNE_ACOLYTE }
 };
 
+Position const SifSpawnPosition = { 2148.301f, -297.8453f, 438.3308f, 2.687807f };
+
+enum Data
+{
+    ACHIEVEMENT_DONT_STAND_IN_THE_LIGHTNING = 29712972,
+    ACHIEVEMENT_SIFFED                      = 29772978,
+    ACHIEVEMENT_LOSE_YOUR_ILLUSION          = 31763183,
+    DATA_CHARGED_PILLAR                     = 1
+};
+
+enum DisplayIds
+{
+    THORIM_WEAPON_DISPLAY_ID                = 45900
+};
+
+uint32 const LightningOrbPathSize = 8;
+G3D::Vector3 const LightningOrbPath[LightningOrbPathSize] =
+{
+    { 2134.889893f, -298.632996f, 438.247467f },
+    { 2134.570068f, -440.317993f, 438.247467f },
+    { 2167.820312f, -440.330261f, 438.247589f },
+    { 2213.394287f, -433.318298f, 412.665863f },
+    { 2227.766113f, -433.275818f, 412.177032f },
+    { 2227.551270f, -263.081085f, 412.176880f },
+    { 2202.208008f, -262.939270f, 412.168976f },
+    { 2182.310059f, -263.233093f, 414.739410f }
+};
+
+Position const ArenaCenter = { 2135.0f, -263.0f, 420.0f, 0.0f }; // used for trash jump calculation
+Position const LightningFieldCenter = { 2135.0f, -312.5f, 438.0f, 0.0f }; // used for lightning field calculation
+
+// p2 start at 5357
+
+class RunicSmashExplosionEvent : public BasicEvent
+{
+    public:
+        RunicSmashExplosionEvent(Creature* owner) : _owner(owner) { }
+
+        bool Execute(uint64 /*eventTime*/, uint32 /*updateTime*/) override
+        {
+            _owner->CastSpell((Unit*)nullptr, SPELL_RUNIC_SMASH);
```

Comment:
Readability mostly, there are many casts like this, it's the equivalent of CreatureAI::DoCastAOE, since this spell has implicit targets and discards any explicit ones at prepare.